### ISSUE TITLE
Ask about auth_query, auth_user

### ISSUE THREAD

**Issue by vinhnd20** [pgbouncer/pgbouncer#967]
I'm learning how to use `auth_query`, `auth_user` without using `userlist.txt`.
Below are my steps:
- Create user `user1`:
``` bash
postgres=# create user user1 password 'user1pass';
CREATE ROLE
```
- Create database db1:
``` bash
postgres=# create database db1 owner user1;
CREATE DATABASE
```

- After that, I create function: 

``` bash
CREATE OR REPLACE FUNCTION user_search(uname TEXT) RETURNS TABLE (usename name, passwd text) as
$$
  SELECT usename, passwd FROM pg_shadow WHERE usename=$1;
$$
LANGUAGE sql SECURITY DEFINER;
```

- And this is my `pgbouncer.ini`:

``` bash
[databases]
db1 = host=localhost dbname=db1 auth_user=user1

[pgbouncer]
listen_addr = 127.0.0.1
listen_port = 16432
auth_type = md5
auth_query = SELECT usename, passwd FROM user_search($1)
```

- But when I try to connect to it, it doesn't work? Did I do something wrong somewhere? Is there any way to get what I want?

``` bash
PGPASSWORD=user1pass psql -h 127.0.0.1 -p 16432 -U user1 db1
psql: error: connection to server at "127.0.0.1", port 16432 failed: FATAL:  password authentication failed
```
Is there a way to avoid using `userlist.txt`?

**Comment by JelteF** [comment]
You need to add `auth_user` to your `pgbouncer.ini.

**Comment by vinhnd20** [comment]
> You need to add `auth_user` to your `pgbouncer.ini.

You mean now I need to add `auth_user` like this or something ?  
[pgbouncer]
listen_addr = 127.0.0.1
listen_port = 16432
auth_user= user1
auth_type = md5
auth_query = SELECT usename, passwd FROM user_search($1)

**Comment by crvt4722** [comment]
I'm also encountering a problem like that

**Comment by JelteF** [comment]
> > You need to add `auth_user` to your `pgbouncer.ini.
> 
> @JelteF You mean now I need to add `auth_user` like this or something ?

Yes

**Comment by JelteF** [comment]
Oh I now see you have set `auth_user` in the database line. That should work as well. Do you have logs from Postgres and PgBouncer during this time.

**Comment by vinhnd20** [comment]
@JelteF 
This is my logs when I use the following command: 
``` bash
vinh@vinh:~/Desktop/Work/Work-1810-2210/pg_bouncer$ PGPASSWORD=user1pass psql -h 127.0.0.1 -p 16432 -U user1 db1
psql: error: connection to server at "127.0.0.1", port 16432 failed: FATAL:  password authentication failed
```

 - Logs:
 ``` bash
.1, tls: OpenSSL 3.0.8 7 Feb 2023
2023-10-24 15:17:09.320 +07 [9646] LOG C-0x557bb2213ed0: db1/user1@127.0.0.1:50926 login attempt: db=db1 user=user1 tls=no
2023-10-24 15:17:09.321 +07 [9646] LOG C-0x557bb2213ed0: db1/user1@127.0.0.1:50926 closing because: password authentication failed (age=0s)
2023-10-24 15:17:09.321 +07 [9646] WARNING C-0x557bb2213ed0: db1/user1@127.0.0.1:50926 pooler error: password authentication failed
```

**Comment by vinhnd20** [comment]
@JelteF  
If now I want to use `userlist.txt`, which user I should put on this file ? `postgres` or `user1` ?

**Comment by eulerto** [comment]
@vinhnd20 see https://github.com/pgbouncer/pgbouncer/issues/975#issuecomment-1785691673 .

**Comment by emelsimsek** [comment]
@vinhnd20  It looks like `auth_user` is the user who runs the` auth_query `for authenticating other users. You can create a dedicated auth_user for this purpose. It does not have to be the username in the connection string.

Now this user should be able to authenticate itself in order to run `auth_query.` Hence you need to specify this user in auth_file.

The improvement is that you only have to have one line in auth_file, instead of listing all the Postgres users.

In summary you still need an auth_file with the auth_user and its password in it.

**Comment by garry-t** [comment]
not so helpful. I also stuck with problem. After I read guide , In my understanding no need any `auth_file` anymore if I specified config like below: 
```
auth_type = md5
auth_user = pgbouncer
auth_dbname = test-db
auth_query = SELECT usename, passwd FROM user_search($1)
```
user `pgbouncer` is db owner. 
For login I use same user
`psql -U pgbouncer -p 6432 -h 127.0.0.1 test-db`
But I getting same error. 
`WARNING C-0x56297658ea70: test-db/pgbouncer@127.0.0.1:39624 pooler error: password authentication failed`
I followed @emelsimsek suggestion. Still doesn't work.

**Comment by garry-t** [comment]
@vinhnd20 did you find solution for yourself ?

**Comment by eulerto** [comment]
@garry-t see my [previous comment](https://github.com/pgbouncer/pgbouncer/issues/967#issuecomment-1785695563). You need the `auth_file`. How will PgBouncer know the credentials to log into Postgres and execute the `auth_query`? `auth_file` should provide the `auth_user` credentials.

**Comment by garry-t** [comment]
Ok, yep. 
So for those who will struggle as me. You need to have following in config: 

```
[pgbouncer]
logfile = /var/log/pgbouncer/pgbouncer.log
pidfile = /run/pgbouncer/pgbouncer.pid
listen_addr = 0.0.0.0
listen_port = 6432
unix_socket_dir = /var/run/pgbouncer
auth_type = md5
auth_user = pgbouncer
auth_dbname = test_db
auth_query = SELECT usename, passwd FROM user_search($1)
auth_file = /etc/pgbouncer/userlist.txt
admin_users = postgres
stats_users = postgres
ignore_startup_parameters = extra_float_digits,geqo,search_path
```
inside `auth_file` I specified user `pgbouncer` and `password`. And off course created search function for this db in config and assigned permissions to `auth_user`. 

Ticket can be closed in my opinion.

**Comment by sxllllll** [comment]
I have tried many ways, but still cannot get rid of userlist.txt

**Comment by terencehonles** [comment]
@sxllllll @artjxm in case you haven't figured it out, I've found a way to do what you're trying to do (or at least specifically what @artjxm is doing).

Using `auth_dbname` you can force the connection type by *also* defining the db in the database section. I took the inspiration from the PR author's blog post https://medium.com/@daymelbonne/pgbouncer-improving-auth-query-authentication-c6c5c441e22 (well at least it looks like they were *one* of the people that tried to implement the functionality)

This is roughly what I have (still playing with config, and considering the auth_query SECURITY DEFINER suggestion):

```ini
[databases]
; for the postgres DB use the connection as is (or specify the path to the unix socket)
postgres =
; include any other per db config you need here...
; for me that's all other DBs connecting via TCP (because my current hba.conf),
* = host=localhost

[pgbouncer]
; use the postgres database definition above.
; Since pg_authid is shared across all databases pointing to a single DB is OK
; (I initially assumed no and why I initially skipped this option, but reading
; the blog post and also looking at the PG docs / testing the auth_query out -- it
; all checks out)
auth_dbname = postgres
; configure PG HBA to ensure `peer` authentication works for the configured `auth_user`
auth_hba_file = /etc/postgresql/pg_hba.conf
auth_ident_file = /etc/postgresql/pg_ident.conf
auth_type = hba
; user who is allowed to login with `peer` authentication
auth_user = postgres
; make sure your socket directory matches your PG version
unix_socket_dir = /var/run/postgresql
```

The key is to make sure that the `auth_user` can use peer authentication, and that will require using a unix socket. I was wondering how to specify an auth user connection string, and it's effectively by specifying a database which _can_ have its own connection string (amongst other parameters)
